### ISSUE TITLE
Add strict failure semantics for explicitly requested VAD, alignment, and punctuation stages

### ISSUE THREAD

**Issue by OldBeLight** [CrispStrobe/CrispASR#311]
## Summary

When auxiliary processing stages are explicitly requested through the CLI, CrispASR may currently continue successfully after those stages fail.

This makes it difficult for applications integrating CrispASR through the CLI to distinguish successful degradation from a failed processing requirement.

Affected stages appear to include:

- VAD selected with `--vad-model` / `-vm`
- External forced alignment selected with `--aligner-model` / `-am`
- Punctuation restoration selected with `--punc-model`

### Explicit VAD

When a VAD model cannot be loaded, some paths return an empty slice list.

`crispasr_compute_audio_slices()` then uses file readability to distinguish “the VAD ran and found no speech” from “the model failed to load.” A readable but invalid/incompatible model can therefore be interpreted as successful VAD with no detected speech.

A missing or unreadable model may instead fall back to fixed/energy-based chunking.

In both cases, the explicitly requested VAD stage can fail without making the command fail.

### External forced alignment

When an explicitly selected aligner fails to load, compute logits, perform alignment, or produce words, transcription can continue with the original ASR segments.

The command may exit successfully and `-ojf` may contain segments without the requested `words[]` data.

### Punctuation restoration

When an explicitly selected punctuation model fails to load, the CLI prints a warning similar to:

```text
warning: failed to load punc model ... — continuing without
```

The command then continues and may exit successfully with unpunctuated output.

## Why this is a problem

CLI integrations may treat the following as required task properties:

- audio was segmented using the user-selected VAD;
- every non-empty transcription segment received word-level timestamps;
- punctuation restoration was applied using the selected model.

A zero exit code and a valid JSON file currently do not prove that these requirements were fulfilled.

Parsing human-readable stderr warnings is brittle, and punctuation success cannot be reliably inferred by checking whether the resulting text happens to contain punctuation.

It is also important to distinguish these valid outcomes:

1. VAD loaded and ran successfully but detected no speech;
2. ASR ran successfully and returned an empty transcript;
3. the explicitly requested VAD, aligner, or punctuation stage failed.

These outcomes should not collapse into the same successful CLI result.

## Requested behavior

Please add a machine-reliable way to require explicitly selected pipeline stages to succeed.

One possible design would be an opt-in flag such as:

```text
--strict-pipeline
```

Alternatively, stage-specific requirements could be exposed, for example:

```text
--require-vad
--require-word-timestamps
--require-punctuation
```

Under strict semantics:

- failure to load or execute an explicitly required VAD model returns a non-zero exit code;
- successful VAD execution that detects no speech remains a successful result;
- failure of an explicitly required aligner returns a non-zero exit code;
- for a non-empty transcript, failure to produce valid word timestamps returns a non-zero exit code;
- failure to load or execute an explicitly required punctuation model returns a non-zero exit code;
- partial output may be retained for diagnostics if desired, but it must not be reported as successful completion;
- the behavior must not depend on parsing human-readable stderr;
- direct local model paths must work without invoking auto-download.

For backward compatibility, the current permissive behavior could remain the default if changing it would be disruptive, while integrations opt into strict mode.

## Example integration

An application may invoke CrispASR once with fully prepared local models:

```bash
crispasr \
  --backend <backend> \
  -m /models/asr.gguf \
  -f /workspace/input.wav \
  -l zh \
  --vad \
  -vm /models/vad.gguf \
  -am /models/aligner.gguf \
  --force-aligner \
  --punc-model /models/punctuation.gguf \
  -ojf \
  -of /workspace/result
```

The application needs the process result to reliably mean:

> Every explicitly required stage completed successfully and the requested output properties are present.

## Suggested acceptance cases

1. Valid VAD runs and finds speech: command succeeds.
2. Valid VAD runs and finds no speech: command succeeds with an explicit no-speech result.
3. Missing, corrupt, incompatible, or unloadable required VAD: command fails.
4. Required aligner cannot load or execute: command fails.
5. Non-empty transcript receives no aligned words: command fails when word timestamps are required.
6. Required punctuation model cannot load or execute: command fails.
7. All requested stages succeed: command exits successfully and `-ojf` contains the expected data.
8. Strict mode never triggers model auto-download when direct local paths are supplied.
9. Existing permissive behavior remains available if backward compatibility
```

**Comment by CrispStrobe** [comment]
Done — added `--strict-pipeline` (plus `--require-vad` / `--require-word-timestamps` / `--require-punctuation`): an explicitly-requested VAD/aligner/punctuation stage that fails to load or produce its output now returns a non-zero exit (VAD=30, word-timestamps=31, punctuation=32; config error=2) and skips writing the output file, instead of degrading silently.
A stage that ran and legitimately found nothing (VAD → no speech) stays a success, direct local model paths are used as-is (no auto-download), and nothing depends on parsing stderr.
The HTTP server enforces the same via `strict_pipeline` / `require_*` form fields, returning HTTP 400 with a JSON error body.
Shipped to `main` — docs in [docs/cli.md](https://github.com/CrispStrobe/CrispASR/blob/main/docs/cli.md) and [docs/server.md](https://github.com/CrispStrobe/CrispASR/blob/main/docs/server.md); it will be in the next release.
Thanks for the precise write-up — your acceptance cases mapped almost 1:1 onto the tests.

**Comment by subof** [comment]
Looks like I can jump in here and add my two cents (or five!), so we don't have to start a whole new thread. As a user of the Subtitle Edition, I ran into an issue where the mandatory subtitle alignment pipeline just isn't giving me the result I expected. So, you load the audio source into the window, then you import a plain text file via the File/Import menu, and then you run the alignment using one of the CrispASR models. But instead of fixing it, it actually creates way more chaos! They say CrispASR only works well on short clips, but I think we're at the threshold of solving this core feature right here. A lot of pre-made subtitles aren't very accurate with their timestamps. Usually, you fix that using the MFA console utility and its dictionaries, but now, with CrispASR, we finally have a chance to make it way easier, right?"

**Comment by CrispStrobe** [comment]
Thanks @subof — good news: this is exactly what CrispASR's **`--align-only`** mode is for, and it needs no MFA-style dictionaries. It takes your existing text (plain `.txt` or an unaligned `.srt`) and *force-aligns* it to the audio, re-timing the cues — it does **not** re-transcribe, so your text is preserved.

**I just verified it end-to-end** on `samples/jfk.wav` (11 s). Feeding a deliberately mis-timed SRT through `--align-only` corrected every cue while keeping the text unchanged:

| cue | imported (wrong) | after `--align-only` |
|----|------------------|----------------------|
| 1 `And so my fellow Americans,` | 00:00.000 → 00:02.000 | **00:00.400 → 00:03.200** |
| 2 `ask not what your country can do for you,` | 00:02.000 → 00:04.000 | **00:03.520 → 00:08.000** |
| 3 `ask what you can do for your country.` | 00:04.000 → 00:06.000 | **00:08.320 → 00:11.040** |

```bash
# re-time an SRT against the audio (cue-preserving):
crispasr --align-only -am auto --auto-download -f audio.wav \
    --text-file yoursubs.srt --align-granularity segment --align-output retimed.srt

# or a plain transcript (one line per cue):
crispasr --align-only -am auto --auto-download -f audio.wav \
    --text-file transcript.txt --align-granularity segment --align-output aligned.srt
```
(Word-level check on the same clip: `And` @0.40 s … `country.` ends @11.04 s ≈ the 11.00 s duration — all 22 words monotonic and on the beat.)

So the CLI does the job cleanly. **Please run that command directly on your file** — it pinpoints where the "chaos" comes from:

- **CLI output is well-timed** → the problem is in how Subtitle Edit drives the alignment — most likely it is running full ASR/transcription rather than `--align-only` with your imported text, so the model's own transcript replaces yours.
- **CLI is also bad** → it's a model/audio/text issue; the usual culprits:
  - **Aligner ↔ audio language mismatch.** `-am auto` guesses; force the matching one, e.g. `-am wav2vec2-aligner-fr` (en/es/fr/it/ja/nl/pl/ru/ua available), or `-am canary-ctc-aligner-q4_k.gguf` (the multilingual CTC aligner used above).
  - **The imported text doesn't match what's actually spoken.** Forced alignment assumes the text *is* the transcript — missing/extra/paraphrased words derail the cue (an aligner can't skip mismatches like ASR does).
  - **Leading silence / music before speech** can drift the first timestamps.
  - **Non-Latin scripts** are auto-romanized for CTC alignment (#252), but worth confirming for your language.

To dig in, please share: the **audio language**, **which CrispASR model** you picked in Subtitle Edit, and a short sample — a few seconds of audio + the **exact text you imported** + the **bad output** — so I can tell quickly whether it's the SE integration or the aligner.

(Heads-up: this is really a forced-alignment topic rather than #311's strict-exit semantics — happy to keep it here, but a dedicated issue with that sample would be easier to track.)
